scala/private/rule_impls.bzl

```diff
@@ -65,136 +62,67 @@ def compile_scala(
-        "{k}: {v}".format(k = k, v = v)
-        # We sort the arguments for input stability and reproducibility
-        for (k, v) in sorted(optional_scalac_args_map.items())
-    ])
-
-    plugins_list = plugins.to_list()
-    plugin_arg = _join_path(plugins_list)
-
-    separator = ctx.configuration.host_path_separator
-    compiler_classpath = _join_path(compiler_classpath_jars.to_list(), separator)
+        plugins = depset(transitive = [plugins, ctx.attr._dependency_analyzer_plugin.files])
 
     toolchain = ctx.toolchains["@io_bazel_rules_scala//scala:toolchain_type"]
+    compiler_classpath_jars = cjars if dependency_info.dependency_mode == "direct" else transitive_compile_jars
+    classpath_resources = getattr(ctx.files, "classpath_resources", [])
     scalacopts = [ctx.expand_location(v, input_plugins) for v in toolchain.scalacopts + in_scalacopts]
     resource_paths = _resource_paths(resources, resource_strip_prefix)
     enable_diagnostics_report = toolchain.enable_diagnostics_report
 
-    scalac_args = """
-Classpath: {cp}
-ClasspathResourceSrcs: {classpath_resource_src}
-Files: {files}
-JarOutput: {out}
-Manifest: {manifest}
-Plugins: {plugin_arg}
-PrintCompileTime: {print_compile_time}
-ExpectJavaOutput: {expect_java_output}
-ResourceTargets: {resource_targets}
-ResourceSources: {resource_sources}
-ResourceJars: {resource_jars}
-ScalacOpts: {scala_opts}
-SourceJars: {srcjars}
-StrictDepsMode: {strict_deps_mode}
-UnusedDependencyCheckerMode: {unused_dependency_checker_mode}
-DependencyTrackingMethod: {dependency_tracking_method}
-StatsfileOutput: {statsfile_output}
-EnableDiagnosticsReport: {enable_diagnostics_report}
-DiagnosticsFile: {diagnostics_output}
-""".format(
-        out = output.path,
-        manifest = manifest.path,
-        # Using ':::' as delimiter because ',' can collide with actual scalac options
-        # https://github.com/bazelbuild/rules_scala/issues/1049
-        scala_opts = ":::".join(scalacopts),
-        print_compile_time = print_compile_time,
-        expect_java_output = expect_java_output,
-        plugin_arg = plugin_arg,
-        cp = compiler_classpath,
-        classpath_resource_src = _join_path(classpath_resources),
-        files = _join_path(sources),
-        srcjars = _join_path(all_srcjars.to_list()),
-        # the resource paths need to be aligned in order
-        resource_targets = ",".join([p[0] for p in resource_paths]),
-        resource_sources = ",".join([p[1] for p in resource_paths]),
-        resource_jars = _join_path(resource_jars),
-        strict_deps_mode = dependency_info.strict_deps_mode,
-        unused_dependency_checker_mode = dependency_info.unused_deps_mode,
-        dependency_tracking_method = dependency_info.dependency_tracking_method,
-        statsfile_output = statsfile.path,
-        enable_diagnostics_report = enable_diagnostics_report,
-        diagnostics_output = diagnosticsfile.path,
-    )
+    args = ctx.actions.args()
```

Comment:
Probably it can. But if I just extract whole args to separate function it will contain same huge parameter list as `compile_scala`. If I extract multiple functions to build args I'll lose visibility how final args file would look like. I'll look what I can do about it.